### PR TITLE
chore: guild default tokens

### PR DESCRIPTION
**Pull request by catngh**
**What does this PR do?**

-  `GET: /configs/tokens?guild_id=` returns global default tokens when `guild_id=""`
- Activity log `guild_id` is nullable to log DM activities

image
image

### CODE REVIEW

File: pkg/repo/guild_user_activity_log/pg.go [hunk 1]
```diff
@@ -16,3 +16,7 @@ func NewPG(db *gorm.DB) Store {
 func (pg *pg) CreateOne(record model.GuildUserActivityLog) error {
 	return pg.db.Create(&record).Error
 }
+
+func (pg *pg) CreateOneNoGuild(record model.GuildUserActivityLog) error {
```
Comment (trkhoi) [inline pkg/repo/guild_user_activity_log/pg.go:20]: Why need this ?
CreateOne can also create a record with null guild id

**Review by trkhoi — CHANGES_REQUESTED** [review]

**Comment by trkhoi** [comment]
please provide flowchart or something related, i dont understand your logic

File: pkg/entities/defi.go [hunk 2]
```diff
@@ -209,17 +209,27 @@ func (e *Entity) InDiscordWalletWithdraw(req request.TransferRequest) (*response
...
-		ChannelID: req.ChannelID,
-		UserID:    req.Sender,
-		Timestamp: time.Now(),
-		Action:    req.TransferType,
-	}); err != nil {
-		err = fmt.Errorf("error create activity log: %v", err)
-		return nil, err
+	if req.GuildID == "" {
+		// log activity
+		if err := e.repo.GuildUserActivityLog.CreateOneNoGuild(model.GuildUserActivityLog{
```
Comment (anhnh12) [inline pkg/entities/defi.go:214]: can we reuse `HandleUserActivities()`?

File: pkg/entities/defi.go [hunk 3]
```diff
@@ -289,21 +299,31 @@ func (e *Entity) InDiscordWalletBalances(guildID, discordID string) (*response.U
...
-	if err != nil {
-		err = fmt.Errorf("failed to get guild %s tokens - err: %v", guildID, err)
-		return nil, err
-	}
-	if len(gTokens) == 0 {
-		if err = e.InitGuildDefaultTokenConfigs(guildID); err != nil {
+	var tokens []model.Token
+
+	if guildID == "" {
+		gTokens, err := e.GetGlobalDefaultToken()
```
Comment (anhnh12) [inline pkg/entities/defi.go:305]: why need a separated entity function here?
we already had `repo.Token.GetDefaultTokens()`

Comment (anhnh12) [inline pkg/entities/defi.go:305]: also update `gTokens, err := e.GetGlobalDefaultToken()` => `tokens, err = .....`

**Review by anhnh12 — CHANGES_REQUESTED** [review]

File: pkg/entities/defi.go (hunk 3 above)
Comment (catngh) [reply pkg/entities/defi.go:305]: i made this function for `configs/tokens/guild_id=""`, here i just reuse it

File: pkg/entities/defi.go (hunk 2 above)
Comment (catngh) [reply pkg/entities/defi.go:214]: `HandleUserActivities()` contains many logics to handle level up, roles, guild configs, here i only want to log activity

File: pkg/entities/defi.go (hunk 3 above)
Comment (anhnh12) [reply pkg/entities/defi.go:305]: should keep logic handler at entity layer, not handler layer

File: pkg/repo/guild_user_activity_log/pg.go (hunk 1 above)
Comment (catngh) [reply pkg/repo/guild_user_activity_log/pg.go:20]: string values cannot be null, it will default to `""` which is an error for foreign keys

Comment (trkhoi) [reply pkg/repo/guild_user_activity_log/pg.go:20]: agree

**Review by trkhoi — APPROVED** [review]
will let anhnh decide logic flow

File: pkg/entities/defi.go [hunk 4]
```diff
@@ -289,21 +299,30 @@ func (e *Entity) InDiscordWalletBalances(guildID, discordID string) (*response.U
...
-	gTokens, err := e.repo.GuildConfigToken.GetByGuildID(guildID)
-	if err != nil {
-		err = fmt.Errorf("failed to get guild %s tokens - err: %v", guildID, err)
-		return nil, err
-	}
-	if len(gTokens) == 0 {
-		if err = e.InitGuildDefaultTokenConfigs(guildID); err != nil {
+	var tokens []model.Token
+
+	if guildID == "" {
```
Comment (anhnh12) [inline pkg/entities/defi.go:304]: try to reuse `e.GetGuildTokens()`

**Review by anhnh12 — CHANGES_REQUESTED** [review]

File: pkg/entities/defi.go (hunk 4 above)
Comment (catngh) [reply pkg/entities/defi.go:304]: fixed

**Review by anhnh12 — APPROVED** [review]

**Pull request merged**